plutus-benchmark/bitwise/src/PlutusBenchmark/Ed25519.hs

```diff
@@ -0,0 +1,147 @@
+{-# LANGUAGE MultiWayIf        #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module PlutusBenchmark.Ed25519 (checkValid) where
+
+import GHC.ByteOrder (ByteOrder (LittleEndian))
+import PlutusBenchmark.SHA512 (sha512)
+import PlutusTx.Prelude hiding (inv)
+
+{-# INLINE checkValid #-}
+checkValid ::
+  BuiltinByteString ->
+  BuiltinByteString ->
+  BuiltinByteString ->
+  Bool
+checkValid sig message pubKey =
+  let r = decodePoint (sliceByteString 0 32 sig)
+      a = decodePoint pubKey
+      s = decodeInt (sliceByteString 32 32 sig)
+      hintArg = encodePoint r <> pubKey <> message
+      h = hint hintArg
+   in scalarMult bPoint s == edwards r (scalarMult a h)
+
+-- Helpers
+
+newtype Point = Point (Integer, Integer)
+
+instance Eq Point where
+  {-# INLINE (==) #-}
+  Point (x1, y1) == Point (x2, y2) = x1 == x2 && y1 == y2
+
+{-# INLINE bPoint #-}
+bPoint :: Point
+bPoint = Point (bx `modulo` q, by `modulo` q)
+
+{-# INLINE decodePoint #-}
+decodePoint :: BuiltinByteString -> Point
+decodePoint bs
+  | odd x /= x_0 = Point (q - x, yInt)
+  | otherwise = Point (x, yInt)
+  where
+    x_0 = readBit bs 7
+    x = xRecover yInt
+    yInt = decodeInt $ clearBit 7 bs
+
+{-# INLINE encodePoint #-}
+encodePoint :: Point -> BuiltinByteString
+encodePoint (Point (x, y)) = result
+  where
+    zeroPos :: Integer
+    zeroPos = 7
+    result = writeBits yBS [zeroPos] [xLSBVal]
+    yBS = integerToByteString LittleEndian 32 y
+    xBS = integerToByteString LittleEndian 32 x
+    xLSBVal = readBit xBS 248
+
+{-# INLINE scalarMult #-}
+scalarMult :: Point -> Integer -> Point
+scalarMult p e =
+  if e == 0
+    then Point (0, 1)
+    else
+      let q' = scalarMult p (e `divide` 2)
```

Comment:
This is because `q` is already defined as a top-level constant further down. The naming is unfortunate, but we basically followed the [reference implementation in Python](https://ed25519.cr.yp.to/python/ed25519.py) closely.

Reply:
Oh, of course!